lib/oga/xml/document.rb

```diff
@@ -57,6 +57,12 @@ def root_node
         self
       end
 
+      # Returns the top level element in the document.
+
```

Comment:
@scotchi 

> Hmm, I hadn't looked into the HTML module, but had assumed there was a parallel class tree there specialized for HTML-specific behavior.

No, there are only some extra classes for the parser and for dealing with HTML entities. Classes such as `XML::Document` work for both XML and HTML documents. `XML::Document` _does_ have the `type` attribute which will return `:html` for HTML documents.

> To me it seems inevitable that some modality will be introduced (outside of just the parser) for the differences between XML and HTML.

This exists in a few places and mostly deals with serialisation/entity conversion.

>  I get that that's not particularly elegant (but then, we're talking about parsing HTML), but leaving out behavior that everyone who's ever used an XML parser will expect because of HTML's quirks seems broken.

Oga doesn't aim to be compatible with other parsers out there, instead it aims to be a more elegant and correct parser. I don't want to complicate Oga any further just because other parsers do so.

In the example of `root_element` it doesn't make sense to change behaviour between XML and HTML documents anyway. What would you have it return for HTML documents, all top-level elements? If so that directly conflicts with the name of the method and really isn't that useful.

> In more concrete terms, this means for users of other libraries:
> 
>     doc << element
> 
> Has to be replaced by this:
> 
>     doc.children.last.children << element
> 
> Or, more verbose, but at least readable:
> 
>     root = doc.children.last
>     root.children << element

These examples are confusing. If `Document#<<` would use `root_element` (which makes even less sense) then this would translate to `doc.children.first.children << element` and *not* `doc.children.last.children << element`.

Either way, to summarise:

* I'm not convinced `Node#<<`, `Document#<<` or something similar is useful as there already is a way of doing this by using `node_or_document.children << ...`
* `Document#root_element` doesn't strike me as very useful either (especially the more I think of it), and its behaviour doesn't work for certain HTML documents. Furthermore there's already `Document#root_node` (and `Node#root_node`) and the similar names may confuse users.

I feel that methods such as the above are better suitable for some oga-extensions Gem of sorts. If they become popular enough I may reconsider adding them to Oga.